### PR TITLE
Fix memory leak

### PR DESCRIPTION
**Pull request by OpenSourceAries**
Summary of the changes (Less than 80 chars)
 - Move event registration out of event callback to avoid memory leak(recursive event registration)

Addresses #bugnumber (in this specific format)

### CODE REVIEW

**Comment by BrennanConroy** [comment]
Are you actually seeing an issue?

This is how reload registration works. When the token is triggered you need to get a new token and register the callback for when the next options change happens.

**Comment by OpenSourceAries** [comment]
Sorry, not seeing related issues.

My English is not good enough to explain, Hope the Demo could explain my thought.

``` 
public class FakeOptionsMonitor
{
    internal event Action<string> _onChange;

    public void SimulationSourceChanged(string newOptions)
    {
        _onChange?.Invoke(newOptions);
    }

    public IDisposable OnChange(Action<string> listener)
    {
        var disposable = new ChangeTrackerDisposable(this, listener);
        _onChange += disposable.OnChange;
        return disposable;
    }
}

internal class ChangeTrackerDisposable : IDisposable
{
    private readonly Action<string> _listener;
    private readonly FakeOptionsMonitor _monitor;

    public ChangeTrackerDisposable(FakeOptionsMonitor monitor, Action<string> listener)
    {
        _listener = listener;
        _monitor = monitor;
    }

    public void OnChange(string name) => _listener.Invoke(name);

    public void Dispose() => _monitor._onChange -= OnChange;
}
```

Main:

```
public static void ReloadFakeOptionsChange(string fakeNewOptions)
{
    Console.WriteLine($"Change: {fakeNewOptions}");
    _optionsReloadToken = _fakeOptionsMonitor.OnChange(ReloadFakeOptionsChange);
}

private static FakeOptionsMonitor _fakeOptionsMonitor;
private static IDisposable _optionsReloadToken;
private static void Main(string[] args)
{
    _fakeOptionsMonitor = new FakeOptionsMonitor();
    ReloadFakeOptionsChange("Initiliaze");

    string input = string.Empty;
    while (!string.IsNullOrEmpty(input = Console.ReadLine()))
    {
        Console.WriteLine($"Input: {input}");
        _fakeOptionsMonitor.SimulationSourceChanged(input);
    }
}
```

Test result:  image

Modiy Main:

```
public static void ReloadFakeOptionsChange(string fakeNewOptions)
{
    Console.WriteLine($"Change: {fakeNewOptions}");
}

private static FakeOptionsMonitor _fakeOptionsMonitor;
private static IDisposable _optionsReloadToken;
private static void Main(string[] args)
{
    _fakeOptionsMonitor = new FakeOptionsMonitor();
    _optionsReloadToken = _fakeOptionsMonitor.OnChange(ReloadFakeOptionsChange);
    ReloadFakeOptionsChange("Initiliaze");

    string input = string.Empty;
    while (!string.IsNullOrEmpty(input = Console.ReadLine()))
    {
        Console.WriteLine($"Input: {input}");
        _fakeOptionsMonitor.SimulationSourceChanged(input);
    }
}
```

Or

```
public static void ReloadFakeOptionsChange(string fakeNewOptions)
{
    Console.WriteLine($"Change: {fakeNewOptions}");
    _optionsReloadToken?.Dispose();
    _optionsReloadToken = _fakeOptionsMonitor.OnChange(ReloadFakeOptionsChange);
}

private static FakeOptionsMonitor _fakeOptionsMonitor;
private static IDisposable _optionsReloadToken;
private static void Main(string[] args)
{
    _fakeOptionsMonitor = new FakeOptionsMonitor();
    ReloadFakeOptionsChange("Initiliaze");

    string input = string.Empty;
    while (!string.IsNullOrEmpty(input = Console.ReadLine()))
    {
        Console.WriteLine($"Input: {input}");
        _fakeOptionsMonitor.SimulationSourceChanged(input);
    }
}
```

Test result: image


It's OK to close this PR if you feel I am wrong or you have better solution.

**Comment by rynowak** [comment]
In your first example, you're relying on `ChangeTrackerDisposable .Dispose` to unhook an event handler, but nothing calls dispose on it. 

Since this line runs each time a change is triggered you're going to end up subscribing N times.
```C#
_optionsReloadToken = _fakeOptionsMonitor.OnChange(ReloadFakeOptionsChange);
```

Each time that line runs, you create a new instance of `ChangeTrackerDisposable` but since you didn't dispose the old one, it's still alive and still listening.

**Comment by OpenSourceAries** [comment]
@rynowak  yes, this is what I want to fix through this PR.

**Comment by OpenSourceAries** [comment]
I think the first example could cover the implementation of _ConsoleLoggerProvider_ before this PR.

**Comment by BrennanConroy** [comment]
I just realized this is using `IOptionsMonitor` and not `ChangeToken`. I believe options monitor will handle the underlying change token so registration only needs to happen once.

File: src/Logging/Logging.Console/src/ConsoleLoggerProvider.cs [hunk 1]
```diff
@@ -23,6 +23,7 @@ public ConsoleLoggerProvider(IOptionsMonitor<ConsoleLoggerOptions> options)
             _options = options;
             _loggers = new ConcurrentDictionary<string, ConsoleLogger>();
 
+            _optionsReloadToken = _options.OnChange(ReloadLoggerOptions);
```
Comment (BrennanConroy) [inline src/Logging/Logging.Console/src/ConsoleLoggerProvider.cs]: Could you move this to be below `ReloadLoggerOptions`?

Comment (OpenSourceAries) [reply src/Logging/Logging.Console/src/ConsoleLoggerProvider.cs]: Yes. This is just to keep consistent with other logger providers(such as [Extensions/src/Logging/Logging.AzureAppServices/src/BatchingLoggerProvider.cs](https://github.com/aspnet/Extensions/blob/master/src/Logging/Logging.AzureAppServices/src/BatchingLoggerProvider.cs))

**Comment by BrennanConroy** [comment]
Thanks for finding and fixing this!

**Pull request merged**